### PR TITLE
changed: make PreconditionerType an enum class

### PR DESCRIPTION
**Pull request by akva2**
this to avoid symbol clashes with the implementations. while at it rename it to Type as Preconditioner::PreconditionerType is redundant.

I ran into this while working in this area and found it a bit annoying so let's improve.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this please

File: opm/simulators/linalg/bda/opencl/Preconditioner.cpp
Comment (bska) [inline opm/simulators/linalg/bda/opencl/Preconditioner.cpp:54]: Would it make sense to replace this with a `switch` instead?  That way, the compiler will warn us if we forget one of the alternatives when introducing a new enumerator.  E.g.,
```C++
switch (type) {
case Type::BILU0:
    return ...
    
case Type::CPR:
    return ...

case Type::BISAI:
    return ...
}

OPM_THROW(std::logic_error, "Unsupported preconditioner type " + std::to_string(static_cast<int>(type)));
```

**Review by bska — APPROVED** [review]
I didn't notice before, but does the `CPR` preconditioner really reverse its constructor arguments to the other two?  The others are `(opencl_ilu_parallel, verbosity)`, but the CPR one is `(verbosity, opencl_ilu_parallel)`.  Weird.

In any case, this looks good to me.  The only thing I might consider doing differently is replacing the `if`&ndash;`else` chain with a `switch` on the preconditioner type.

File: opm/simulators/linalg/bda/opencl/Preconditioner.cpp
Comment (akva2) [reply opm/simulators/linalg/bda/opencl/Preconditioner.cpp]: I feared i would fall in that rabbit hole haha. Working in this code I'm constantly fighting the urge to inflate the commits with all sorts of cosmetics and formatting things. Anyways since you called these out; yes the switch makes sense, done. The inconsistent order is indeed inconsistent. I've take care of that as well.

**Comment by akva2** [comment]
jenkins build this please

File: opm/simulators/linalg/bda/opencl/Preconditioner.cpp
Comment (bska) [reply opm/simulators/linalg/bda/opencl/Preconditioner.cpp]: > Working in this code I'm constantly fighting the urge to inflate the commits with all sorts of cosmetics and formatting things.

Right.  If you're doing multiple steps, then we can hold off on those kinds of things until everything else has landed.  Just let me know and then I'll not make those kinds of comments.

Comment (akva2) [reply opm/simulators/linalg/bda/opencl/Preconditioner.cpp]: Oh, please do make suggestion, it only makes the code better.. Don't read it as annoyance with those. It's just the BDA code is really out of wack when it comes to formatting and general little niggles. Too few eyes..

**Review by bska — APPROVED** [review]
Thanks a lot for the updates.  This looks good to me now and I'll merge into master.

**Pull request merged**
